### PR TITLE
SY-2454 - Refactored Color Utilities to Use Functional Patterns

### PR DESCRIPTION
**Pull request by emilbon99**
# Issue Pull Request

## Key Information

- **Linear Issue**: [SY-2454](https://linear.app/synnax/issue/SY-2544)

## Description

The `Color.Color` class has always been problematic because we it makes zod parsing more difficult (check for instanceof) and we often end up putting non-serializable class instances in state. Created a new `color` package in `x/ts` that refactors colors to use the same functional style that we have for the `spatial` packages.

## Basic Readiness

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes to CI.
- [x] I have added needed QA steps to the [release candidate](/synnaxlabs/synnax/blob/main/.github/PULL_REQUEST_TEMPLATE/rc.md) template that cover these changes.
- [x] I have updated in-code documentation to reflect the changes.
- [x] I have updated user-facing documentation to reflect the changes.

### Data Structures

I have ensured that previous versions of stored data structures are properly migrated to new formats in the following projects:

- [x] Server
- [x] Console

### API Changes

The following projects have backwards-compatible APIs:

- [ ] Python Client
- [ ] Server
- [ ] TypeScript Client

### CODE REVIEW

File: x/ts/src/color/color.ts [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+export const hsla = (color: Crude): HSLA => rgbaToHSLA(construct(color));
+
+/**
+ * Creates a new color with the given alpha.
+ *
+ * @param alpha - The alpha value to set. If the value is greater than 1, it will be
+ * divided by 100.
+ * @returns A new color with the given alpha.
+ */
+export const setAlpha = (color: Crude, alpha: number): Color => {
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:165]: test

File: x/ts/src/color/color.spec.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      const expected = [122 / 255, 44 / 255, 38 / 255, 1];
+      expected.forEach((v, i) => {
+        expect(color.rgba1(c)[i]).toBeCloseTo(v);
+      });
+    });
+  });
+  describe("luminance", () => {
+    const tests: Array<[string, number]> = [
+      ["#000000", 0],
+      ["#ffffff", 1],
```
Comment (pjdotson) [inline x/ts/src/color/color.spec.ts:106]: add more tests than just black and white

File: x/ts/src/color/color.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+  if (alpha > 100)
+    throw new Error(`Color opacity must be between 0 and 100, got ${alpha}`);
+  if (alpha > 1) alpha /= 100;
+  return [r, g, b, alpha] as Color;
+};
+
+/**
+ * @returns the luminance of the color, between 0 and 1.
+ */
+export const luminance = (color: Crude): number => {
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:176]: Should add a link showing where this equation is from

File: x/ts/src/color/color.ts [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+  const [r, g, b] = rgb1(color).map((v) =>
+    v <= 0.03928 ? v / 12.92 : ((v + 0.055) / 1.055) ** 2.4,
+  );
+  return Number((0.2126 * r + 0.7152 * g + 0.0722 * b).toFixed(3));
+};
+
+/**
+ * @returns the grayness of the color, between 0 and 1.
+ */
+export const grayness = (color: Crude): number => {
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:186]: Should explain this is a measure we made up ourselves

File: x/ts/src/color/color.ts [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  return 1 - deviation;
+};
+
+/**
+ * @returns the contrast ratio between this color and the given color. The contrast
+ * ratio is a number between 1 and 21, where 1 is the lowest contrast and 21 is the
+ * highest.
+ * @param other
+ * @returns
+ */
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:198]: Fix this documentation. Also add a link to a source for this formula

File: x/ts/src/color/color.spec.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      ["#ffffff", "#ffffff", 1],
+    ];
+    tests.forEach(([hex1, hex2, expected]) => {
+      test(`${hex1} ${hex2}`, () => {
+        const c1 = color.construct(hex1);
+        const c2 = color.construct(hex2);
+        expect(color.contrast(c1, c2)).toBeCloseTo(expected);
+      });
+    });
+    test("pick c with highest contrast", () => {
```
Comment (pjdotson) [inline x/ts/src/color/color.spec.ts:129]: Add a few more tests for pickByContrast on edge cases

File: x/ts/src/color/color.ts [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+/**
+ * @returns the color with the highest contrast ratio to the given colors.
+ */
+export const pickByContrast = (source: Crude, ...colors: Crude[]): Color => {
+  if (colors.length === 0)
+    throw new Error("[Color.pickByContrast] - must provide at least one color");
+  const source_ = construct(source);
+  const [best] = colors
+    .map((c) => construct(c))
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:215]:
```suggestion
    .map(construct)
```

File: x/ts/src/color/color.ts [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+    .sort((a, b) => contrast(source_, b) - contrast(source_, a));
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:216]: Don't really like doing a sorting function here, it makes more sense to just loop
through `colors` and store the highest number.

File: x/ts/src/color/color.ts [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+/**
+ * @returns true if the color is dark.
+ */
+export const isDark = (color: Crude): boolean => luminance(color) < 0.5;
+
+/**
+ * @returns true if the color is light.
+ */
+export const isLight = (color: Crude): boolean => !isDark(color);
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:228]: If the luminance is 0.5 it should be neither `isLight` or `isDark` right? and add test

File: x/ts/src/color/color.ts [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+  const [best] = colors
+    .map((c) => construct(c))
+    .sort((a, b) => contrast(source_, b) - contrast(source_, a));
+  return best;
+};
+
+/**
+ * @returns true if the color is dark.
+ */
+export const isDark = (color: Crude): boolean => luminance(color) < 0.5;
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:223]: add test

File: x/ts/src/color/color.ts [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+
+/**
+ * @returns true if the color is light.
+ */
+export const isLight = (color: Crude): boolean => !isDark(color);
+
+export const colorZ = z
+  .union([hexZ, rgbaZ, rgbZ])
+  .transform((v) => construct(v as string));
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:232]: No type assertion here please

File: x/ts/src/color/color.ts [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+};
+
+/** A totally zero color with no alpha. */
+export const ZERO: Color = [0, 0, 0, 0];
+
+const rgbaToHex = (n: number): string => Math.floor(n).toString(16).padStart(2, "0");
+const hexToRgba = (s: string, n: number): number => parseInt(s.slice(n, n + 2), 16);
+const stripHash = (hex: string): string => (hex.startsWith("#") ? hex.slice(1) : hex);
+
+export const fromHSLA = (hsla: RGBA): RGBA => {
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:253]: add comment + test + input type should be hsla.

File: x/ts/src/color/color.ts [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+  let [h, s, l] = hsla;
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:254]: You should parse that this is a valid HSLA first

File: x/ts/src/color/color.ts [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+  const a = hsla[3];
+  h /= 360;
+  s /= 100;
+  l /= 100;
+  let r, g, b;
+
+  if (s === 0)
+    r = g = b = l; // achromatic
+  else {
+    const q = l < 0.5 ? l * (1 + s) : l + s - l * s;
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:264]: should link to source for these equations

File: pluto/src/vis/diagram/edge/paths.tsx [hunk 15]
```diff
@@ -7,58 +7,65 @@
...
+  color?: color.Crude;
 }
 
-const Pipe = ({ points, color, ...rest }: PathProps): ReactElement => (
+const Pipe = ({ points, color: pathColor, ...rest }: PathProps): ReactElement => (
   <BaseEdge
     path={calcPath(points)}
     style={{
-      stroke: Color.cssString(color),
+      stroke: color.hex(pathColor),
```
Comment (pjdotson) [inline pluto/src/vis/diagram/edge/paths.tsx:28]: These are all `color.hex` instead of `color.cssString` in this file = was this intentional?

File: pluto/src/vis/measure/aether/measure.ts [hunk 16]
```diff
@@ -217,13 +219,29 @@ export class Measure extends aether.Leaf<typeof measureStateZ, InternalState> {
...
+      fill: color.setAlpha(twoValue.color, 0.5),
+      radius: 8,
+      position: twoPos,
+    });
+    draw.circle({
+      fill: color.setAlpha(twoValue.color, 0.8),
+      radius: 5,
+      position: twoPos,
+    });
+    draw.circle({ fill: color.construct("#ffffff"), radius: 2, position: twoPos });
```
Comment (pjdotson) [inline pluto/src/vis/measure/aether/measure.ts:244]: would be good not to have to construct this color every time (make it a constant), or better yet export a
BLACK and WHITE from the x colors package

File: console/src/layout/types/v0.ts [hunk 17]
```diff
@@ -141,7 +141,7 @@ export interface State<A = unknown> {
 
 const themeZ = Theming.themeZ.transform(
```
Comment (pjdotson) [inline console/src/layout/types/v0.ts:142]: we shouldn't need our own themeZ in console now that we don't store a class inside Theme

File: console/src/layout/types/v4.ts [hunk 18]
```diff
@@ -18,7 +18,7 @@ export const VERSION = "4.0.0";
 
 export const sliceStateZ = v3.sliceStateZ.omit({ version: true }).extend({
   version: z.literal(VERSION),
-  colorContext: Color.contextStateZ.transform(Color.transformColorsToHex),
+  colorContext: Color.contextStateZ.transform(color.transformColorsToHex),
```
Comment (pjdotson) [inline console/src/layout/types/v4.ts:21]: should be able to get rid of this now

File: console/src/layout/types/v4.ts [hunk 19]
```diff
@@ -35,6 +35,6 @@ export const sliceMigration: migrate.Migration<v3.SliceState, SliceState> =
     migrate: (s) => ({
       ...s,
       version: VERSION,
-      colorContext: Color.transformColorsToHex(Color.ZERO_CONTEXT_STATE),
+      colorContext: color.transformColorsToHex(Color.ZERO_CONTEXT_STATE),
```
Comment (pjdotson) [inline console/src/layout/types/v4.ts:38]: should be able to remove this

File: x/ts/src/color/color.ts [hunk 20]
```diff
@@ -0,0 +1,319 @@
...
+    const p = 2 * l - q;
+    r = hueToRgb(p, q, h + 1 / 3);
+    g = hueToRgb(p, q, h);
+    b = hueToRgb(p, q, h - 1 / 3);
+  }
+
+  return [Math.round(r * 255), Math.round(g * 255), Math.round(b * 255), a];
+};
+
+const hueToRgb = (p: number, q: number, t: number): number => {
```
Comment (pjdotson) [inline x/ts/src/color/color.ts:274]: rename this to `hueToRGB`

**Review by pjdotson — CHANGES_REQUESTED** [review]
Mostly looks good and am glad for the change.

Two pieces of code that need to get changed:

1. `console/src/layout/selectors.ts` has a function `selectColorContext` that can now be simplified.
2. ctrl+f for`@ts-expect-error`s that are now out of date.

File: pluto/src/text/Text.tsx [hunk 21]
```diff
@@ -58,12 +58,12 @@ export const Text = <L extends text.Level = text.Level>({
...
-  color?: Color.Crude | boolean,
+  colorVal?: color.Crude | boolean,
   shade?: number,
 ): string | undefined => {
-  if (color != null) {
-    if (typeof color === "boolean") return undefined;
-    return Color.cssString(color) as string;
+  if (colorVal != null) {
+    if (typeof colorVal === "boolean") return undefined;
+    return color.cssString(colorVal);
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply pluto/src/text/Text.tsx:66]: This makes it easy to do conditionals in props. i.e. `myProp == true && color`. What I'm going to do is change it to `false`

File: x/ts/src/color/transformColorsToHex.ts [hunk 22]
```diff
@@ -18,8 +18,9 @@ export type ReplaceColorWithHex<T> = T extends Color
       : T;
 
 export const transformColorsToHex = <T>(obj: T): ReplaceColorWithHex<T> => {
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/transformColorsToHex.ts:20]: Removed. Are there any migration checks we should do for this?

File: x/ts/src/color/palette.ts [hunk 23]
```diff
@@ -9,12 +9,12 @@
...
+import { colorZ } from "@/color/color";
 
 const styleZ = z.object({
   key: z.string(),
   name: z.string(),
-  color: Color.z,
+  color: colorZ,
 });
 
 export type Style = z.infer<typeof styleZ>;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/palette.ts:20]: I mean they are used as part of each color palette

File: x/ts/src/color/palette.ts
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/palette.ts:1]: Question here though is what if we ever want to save a color palette into the cluster? We'd need that schema in the client ya know

File: x/ts/src/color/gradient.ts [hunk 24]
```diff
@@ -9,7 +9,16 @@
...
+  bValue,
+  type Color,
+  construct,
+  crudeZ,
+  gValue,
+  rValue,
+  ZERO,
+} from "@/color/color";
 
 export const stopZ = z.object({
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/gradient.ts:23]: We do use it in the color gradient selector. We can not export it? But if we want to remove it should be as part of a larger refactor of the gradient selector

File: x/ts/src/color/gradient.ts [hunk 25]
```diff
@@ -28,36 +37,42 @@ export const fromGradient = (gradient: Gradient, position: number): Color => {
   if (gradient.length === 0) return ZERO;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/gradient.ts:37]: I think all gradient related work should be addressed in a separate issue I linked to above. This will likely need an implementation and even interface refactor.

File: x/ts/src/color/color.ts [hunk 26]
```diff
@@ -0,0 +1,319 @@
...
+  colorZ.safeParse(color).success;
+
+export const isColor = (color: unknown): color is Color =>
+  rgbaZ.safeParse(color).success;
+
+/**
+ * An unparsed representation of a color i.e. a value that can be converted into
+ * a Color object.
+ */
+export type Crude = Hex | RGBA | Color | string | RGB;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts:34]: No need for a string you are right, but I feel like it makes it clearer in the code to understand what we accept as a color to define the type explicitly. Otherwise you'll have to do a bunch of go to definitions to understand what formats we accept

File: x/ts/src/color/color.ts [hunk 27]
```diff
@@ -0,0 +1,319 @@
...
+    hexToRgba(hex_, 2),
+    hexToRgba(hex_, 4),
+    hex_.length === 8 ? hexToRgba(hex_, 6) / 255 : alpha,
+  ];
+};
+
+/** A totally zero color with no alpha. */
+export const ZERO: Color = [0, 0, 0, 0];
+
+const rgbaToHex = (n: number): string => Math.floor(n).toString(16).padStart(2, "0");
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts:250]: Because it doesn't matter what you do. 245.8 going to 245 is not going to make a difference to the user. Want me to round?

File: pluto/src/css/color.ts [hunk 28]
```diff
@@ -7,7 +7,7 @@
 // License, use of this software will be governed by the Apache License, Version 2.0,
 // included in the file licenses/APL.txt.
 
-import { color } from "@/color/core";
+import { color as core } from "@synnaxlabs/x";
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply pluto/src/css/color.ts]: 3/10

File: x/ts/src/color/color.ts [hunk 29]
```diff
@@ -0,0 +1,319 @@
...
+const hexRegex = /^#?([0-9a-f]{6}|[0-9a-f]{8})$/i;
+const hexZ = z.string().regex(hexRegex);
+const rgbValueZ = z.number().min(0).max(255);
+const alphaZ = z.number().min(0).max(1);
+const rgbaZ = z.tuple([rgbValueZ, rgbValueZ, rgbValueZ, alphaZ]);
+const rgbZ = z.tuple([rgbValueZ, rgbValueZ, rgbValueZ]);
+
+export type RGBA = [number, number, number, number];
+export type HSLA = [number, number, number, number];
+export type RGB = [number, number, number];
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts:25]: 2/10

File: x/ts/src/color/color.ts [hunk 30]
```diff
@@ -0,0 +1,319 @@
...
+
+  // Convert hue to degrees
+  h *= 360;
+  s *= 100;
+  l *= 100;
+
+  return [Math.round(h), Math.round(s), Math.round(l), a];
+};
+
+export const crudeZ = z.union([hexZ, rgbaZ, z.string(), rgbZ]);
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts]: `5/10`

File: x/ts/src/color/color.ts [hunk 31]
```diff
@@ -0,0 +1,319 @@
...
+const alphaZ = z.number().min(0).max(1);
+const rgbaZ = z.tuple([rgbValueZ, rgbValueZ, rgbValueZ, alphaZ]);
+const rgbZ = z.tuple([rgbValueZ, rgbValueZ, rgbValueZ]);
+
+export type RGBA = [number, number, number, number];
+export type HSLA = [number, number, number, number];
+export type RGB = [number, number, number];
+export type Hex = z.infer<typeof hexZ>;
+
+export const isCrude = (color: unknown): color is Crude =>
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts:34]: `8/10`

File: x/ts/src/color/color.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts]: 10/10

File: x/ts/src/color/color.ts [hunk 32]
```diff
@@ -0,0 +1,319 @@
...
+ * @param alpha - The alpha value to set. If the value is greater than 1, it will be
+ * divided by 100.
+ * @returns A new color with the given alpha.
+ */
+export const setAlpha = (color: Crude, alpha: number): Color => {
+  const [r, g, b] = construct(color);
+  if (alpha > 100)
+    throw new Error(`Color opacity must be between 0 and 100, got ${alpha}`);
+  if (alpha > 1) alpha /= 100;
+  return [r, g, b, alpha] as Color;
```
[the remark this answers is not in the document]
Comment (emilbon99) [reply x/ts/src/color/color.ts:177]: 2/10

File: x/ts/src/color/color.ts (hunk 14 above)
Comment (emilbon99) [reply x/ts/src/color/color.ts:266]: I don't have a source. This is GPT generated

**Review by pjdotson — APPROVED** [review]
